### DISCUSSION TITLE
[RFC] New Component: Message

### DISCUSSION THREAD

**Discussion by YumoImer** [ant-design/x#16] [RFC]
## Summary
Message 组件是在对话中用于承载用户输出内容的容器
![image](https://github.com/ant-design/x/assets/98121322/b8748ccd-66d9-4afd-803c-10f622362d51)

## Motivation
无论是传统的 IM 应用或者是当下各类大模型对话应用，消息是最核心的组成部分。通常承载的内容可以分两类：
 - 文本类：纯文本、Markdown 格式文本、HTML 格式文本、LaTeX格式文本等
 - 媒体类：图片、音频、视频、文件等

无论哪种消息类型，都可以将布局、通用交互进行抽象，以减少用户的开发、使用成本。另外，消息内容可以考虑支持行业内常见的几种类型，如文、图、音、视、文件等。
### Message 消息组件
``` typescript
export type MesssageSlotType = 'avatar' | 'header' | 'footer' | 'actions';

export interface MessageProps {
  /** 
   * @desc 用于展示消息内容的区域
   * @descEN Area to display the message content
   * @default undefined
   */
  children?: React.ReactNode;

  /**
   * @desc 用于控制消息的内部布局
   * @descEN Controls the internal layout of the message
   * @default 'default'
   */
  layout?: 'default' | 'reverse';

  /** 
   * @desc 用于展示用户头像的区域
   * @descEN Area to display the user avatar
   * @default undefined
   */
  avatar?: React.ReactNode;

  /** 
   * @desc 用于在消息顶部进行扩展的区域
   * @descEN Area to extend at the top of the message
   * @default undefined
   */
  header?: React.ReactNode;

  /** 
   * @desc 用于在消息底部进行扩展的区域
   * @descEN Area to extend at the bottom of the message
   * @default undefined
   */
  footer?: React.ReactNode;

  /** 
   * @desc 用于在消息旁边进行扩展的区域
   * @descEN Area to extend beside the message
   * @default undefined
   */
  actions?: React.ReactNode;

  /** 
   * @desc 语义化结构 style
   * @descEN Semantic structure styles
   * @default undefined
   */
  styles?: Record<MesssageSlotType | 'children', React.CSSProperties>;

  /** 
   * @desc 语义化结构 className
   * @descEN Semantic structure classNames
   * @default undefined
   */
  classNames?: Record<MesssageSlotType | 'children', string>;

  /** 
   * @desc 用于控制是否在加载中状态
   * @descEN Controls whether it is in loading state
   * @default false
   */
  loading?: boolean;
}
```
### 基础使用
``` tsx
import { Message } from '@ant-design/x';

() => (
  <Message>Hello World!</Message>
);
```
### 展示方向
``` tsx
import { Message } from '@ant-design/x';

() => (
  <>
    <Message
      avatar={<div>avatar</div>}
     actions={<div>actions</div>}
    />
      Hello World!
    </Message>
        
    <Message
      layout="reverse"
      avatar={<div>avatar</div>}
      actions={<div>actions</div>}
    />
      Hello World!
    </Message>
  </>
)
```
### 输入中状态
``` tsx
import { Message } from '@ant-design/x';

() => (
  <Message loading />
    Hello World!
  </Message>
)
```
### 扩展区域
``` tsx
import { Message } from '@ant-design/x';

() => (
  <Message
  avatar={<div>avatar</div>}
  header={<div>header</div>}
  footer={<div>footer</div>}
  actions={<div>actions</div>}
/>
  Hello World!
</Message>
)
```
### Design Token 使用
基础文本色：控制内容区域「children」文字颜色
基础背景色：控制内容区域「children」背景颜色
文字尺寸：控制内容区域「children」文字尺寸
圆角：控制内容区域「children」圆角

**Comment by afc163** [comment]
对比一下 https://github.com/ant-design/ant-design/pull/48384 这一版的 API 设计？看看有没有融合的

**Reply by YumoImer** [reply]
对比了下：
- 对于展示方向，两个 RFC 不一致`direction: 'left' | 'right'`、 `placement: 'start' | 'end'`
- 对于加载中状态，两个 RFC 不一致 `loading`、`typing`
- 对于消息内容传入，两个 RFC 不一致 `children`、`content`
- 自定义消息内容，两个 RFC 不一致 `children`、`contentRender`

消息展示方向，背后的实现是 `flex-direction`。或者可以这样：`reverse: boolean`
加载中状态，下面回复了
消息内容，个人感觉还是使用 `children` 好些，内容的渲染交由使用者决定

**Reply by afc163** [reply]
direction 是参考了 css 属性的用词，reverse 感觉更像是纵向的时间顺序？

**Reply by afc163** [reply]
>  消息内容，个人感觉还是使用 children 好些，内容的渲染交由使用者决定

我也同意是 children，更符合直觉。

**Reply by YumoImer** [reply]
> direction 是参考了 css 属性的用词，reverse 感觉更像是纵向的时间顺序？

这个属性本质的用途含义是：决定消息从左到右展示还是从右到左展示。

然后根据这个本质含义，我觉得`direction: 'left' | 'right'`这个方案是可以否了。

然后就是`placement: 'start' | 'end'`，这个 key 之前被用于像 Tooltip 等组件，代表的是相对位置，感觉用在这里不太合适，因为本质是决定一个展示方向。

刚找了一圈 antd 的现存 API，发现还有个方案：`align: 'start' | 'end'` ,感觉更符合语义一些，后续还可以扩展为`align: 'start' | 'end' | 'center'`，居中对齐也有一些使用场景：系统提示消息，如时间、提示消息等。

**Reply by arvinxx** [reply]
左到右这种不合适的，要考虑 RTL 的用户

**Reply by YumoImer** [reply]
> 左到右这种不合适的，要考虑 RTL 的用户

嗯嗯，左右确实不合适，或者说本质的用途是：决定消息本身的布局，头像 | 消息 | 扩展  或者 扩展 | 消息 | 头像

**Reply by ONLY-yours** [reply]
sider 感觉不需要？在左右回复的情况下对这个 sider 可能会有歧义，开一个 contentRender 之类的内容区域渲染出去就够了吧

**Reply by YumoImer** [reply]
> sider 感觉不需要？在左右回复的情况下对这个 sider 可能会有歧义，开一个 contentRender 之类的内容区域渲染出去就够了吧

`sider` 的渲染我想的是在消息内容区域之外，比如撤回、引用、重新生成、编辑、点赞等操作。

消息内容区域之内的，可以使用 `children` 属性

**Reply by arvinxx** [reply]
> sider 的渲染我想的是在消息内容区域之外，比如撤回、引用、重新生成、编辑、点赞等操作。

叫 sider 不一定符合语义，比如钉钉的这种

<img width="529" alt="image" src="https://github.com/ant-design/x/assets/28616219/dfd795a8-3db6-4e63-9ad8-66bf91b51ca3">

Discord 的这种

<img width="661" alt="image" src="https://github.com/ant-design/x/assets/28616219/24799d0b-94fe-421f-9d5c-1ad59ca1527c">

NextChat 的这种

<img width="893" alt="image" src="https://github.com/ant-design/x/assets/28616219/94789ff9-82ae-4431-9a0a-c48d537625dd">


都是操作部分，钉钉的是在 side，但是 Discord 是在右上角，NextChat 甚至在左上角…

从功能语义上抽象来看，我感觉叫 action 可能更合适？

**Reply by afc163** [reply]
actions

**Reply by YumoImer** [reply]
我们需不需要考虑 actions 的位置，感觉这个应该由设计师的设计规范定义

**Reply by afc163** [reply]
从语义上出发确实 sider 这个命名会限制位置。

**Reply by arvinxx** [reply]
> 我们需不需要考虑 actions 的位置，感觉这个应该由设计师的设计规范定义

组件可以有个默认的位置，按设计规范来就好。至于自定义的部分交由 classNames ，让组件消费者自行改位置。甚至我们可以有一些action 放在其他位置的演示 demo 给到用户。

**Reply by YumoImer** [reply]
对于前 `direction: 'left' | 'right'`属性，考虑其语义上代表：消息组件的内部布局而不是消息组件的相对位置。

故修改为`layout?: 'default' | 'reverse'`：
 - 默认值 `default`:代表「头像 | 消息 | 扩展」布局
 - `reverse` 代表「扩展 | 消息 | 头像」布局

**Comment by afc163** [comment]
有 loading 么？

**Reply by YumoImer** [reply]
我这样想的：typing -> loading / done 

done 在一些模型流式返回的数据结构里比较常见

这样可以在语义上兼容文本、图片的流式非流式。都比 typing 好些感觉

**Reply by afc163** [reply]
done 是啥意思？有图么

**Reply by YumoImer** [reply]
> done 是啥意思？有图么

ollama 的 chat 接口返回结构中使用 `done: boolean`作为是否完成的标识：https://github.com/ollama/ollama/blob/main/docs/api.md#response-8

ChatGPT 的 接口返回结构使用 `status: 'in_progress' | 'finished_successfully'`

其他的还有用像 `is_finished: boolean` 这样的

**Reply by arvinxx** [reply]
两个状态不一样吧，一个是当前消息是否是 loading 态，另外一个是消息内容是否正在输出中。

前者存在的场景，比如我发起创建请求，但是服务端还没返回成功响应的情况下（乐观更新中非常常见）。

后者才是内容输出是否完成的标记

**Reply by YumoImer** [reply]
> 两个状态不一样吧，一个是当前消息是否是 loading 态，另外一个是消息内容是否正在输出中。
> 
> 前者存在的场景，比如我发起创建请求，但是服务端还没返回成功响应的情况下（乐观更新中非常常见）。
> 
> 后者才是内容输出是否完成的标记

我的想法是这样的：发起请求后 pending 的过程更多的交互在发送按钮上，消息区域触发渲染的大多数情况是前置请求调用已经成功，然后消息这里只关注是否完成。还有一种情况是请求中途暴毙，这种情况带有点业务属性，感觉我们最好不要考虑进去。所以感觉一个属性 key 做这个事情就好了。

**Reply by arvinxx** [reply]
也行， Message 作为原子组件的确是可以不用考虑这个问题，单纯有个输出中的就好了。

**Comment by afc163** [comment]
可以加上 classNames 和 styles，类似 antd 组件现在的最佳实践，方便用户自定义。

**Reply by YumoImer** [reply]
👌
